integration/resources/templates/combination/graphqlapi_pipeline_resolver.yaml

```diff
@@ -85,7 +85,7 @@ Resources:
             version: Int!
         }
       Auth:
-        Type: API_KEY
+      - Type: API_KEY
```

Comment:
From summary:

> The first item in the list will be the primary authentication, and everything following it maps to additional authorizers.

What does this mean? Specifically "primary authentication".

What is the implication of that? Does a customer need to know that the first thing is "primary" and its implications (if any)?

Reply:
In AppSync, customers can describe numerous authentication providers. They have a primary authentication method which they define, and then they have a list of additional authentication providers. [In CFN](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-appsync-graphqlapi.html), you can see that there is an `AuthenticationType` property and then various config properties to go with that. There is also  a property `AdditionalAuthenticationProviders` where you can list more of them.

Primary authentication is the default authentication method for all fields in a GraphQLSchema.
With additional authentication methods, AppSync allows you to mark your GraphQLSchema with headers to specify different authentication methods for that field. [This section is well documented here in AppSync docs.](https://docs.aws.amazon.com/appsync/latest/devguide/security-authz.html#:~:text=us%2Dwest%2D2%22%7D%27-,Using%20additional%20authorization%20modes,-When%20you%20add)

AppSync customers will already be familiar with this concept of primary vs additional authentication, this just makes how they define them more consistent IMO.

Reply:
> AppSync customers will already be familiar with this concept of primary vs additional authentication

But are they familiar with the concept that the first item in an array? That's my main concern, since it's implicit from the template (there's nothing indicating it's somehow special), and customer must be aware of that behavior.

Reply:
I see what you're asking now. No, this would be a new behaviour that does not currently exist in the CFN shape. I think it makes intuitive sense though (lists are ordered data structures), and should be easy to describe in documentation.

Reply:
What if we make it something like:

```yaml
Auth:
  Primary: API_KEY
  Secondary:
    - FooBar
    - SomethingElse
```

I don't agree it's intuitive in its current state; it's not an existing pattern, and there's nothing even suggesting the first item implies some special behavior. The customer would have to be intimately familiar with it, and learn to notice this specific property, which is a non-negligible cognitive ask considering templates can easily span 1000+ lines.

Reply:
Also `AuthenticationType` is a string, while `AdditionalAuthenticationProviders` is a [list of objects](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appsync-graphqlapi-additionalauthenticationprovider.html). In the current implementation does it mean first item in the list must be a string, and the rest must also be objects?

Reply:
No. The objects in `AdditionalAuthenticationProviders` is basically just the `AuthenticationType` property and a `Config` property if the type of authentication requires it.

Our implementation makes it so every item in the list has the shape of this object, `AuthenticationType` and `xyzConfig` if necessary.

Reply:
In terms of AppSync, those actually are `Default` and `Additional` instead of Primary and Secondary. And to keep it is as familiar for customers as possible, we can just keep it exactly as it is in AppSync without even additional `Auth` level. I liked the idea of having it as a list, as it seemed to be better organized and I thought that naturally a customer adds a single auth provider and it becomes a default one and then they append to the array one or more providers. But... somebody will obviously insert a new provider in the front of the array, because, 'why not?' and 'nobody likes reading docs' and breaks API. So yes, the safest thing is to keep it as in AppSync